### PR TITLE
remove duplicate fcntl with O_NONBLOCK call

### PR DESCRIPTION
**Pull request by chenbd**
try fixing issues #1535

### CODE REVIEW

**Comment by chenbd** [comment]
yes, ti's  incomplete.
will dig into this and try to remove the fcntl() call inside create_socket().

**Comment by kazuho** [comment]
@volyrique 
> IMHO the underlying issue is the fcntl() call inside create_socket() (on line 386), so you should try removing it, and adjusting all callers. Doing it the other way round (as you have started) means missing on the useful optimization provided by accept4(), for example.

👍

**Comment by kazuho** [comment]
@chenbd How about setting the `O_NONBLOCK` flag in the public functions of evloop.c.h (e.g., `h2o_evloop_socket_create`, `h2o_evloop_socket_accept`, `h2o_socket_connect`)?

The reason I suggest this is because some functions that call the socket API (e.g. `cloexec_pipe`) need to return a blocking socket.

**Comment by chenbd** [comment]
@kazuho 
Do we need return a blocking socket (e.g. cloexec_pipe).
guess all socket/pipe should be in non-blocking mode? please correct me if anything wrong.

we can benefit from binding close-exec and non-block flags with single call under linux.

**Comment by kazuho** [comment]
> Do we need return a blocking socket (e.g. cloexec_pipe).

Yes. The assumption is that file descriptors registered to evloop are non-blocking. But there are descriptors that we do not register to evloop. Such sockets are handled as blocking sockets in their own thread. Examples include OCSP stapling, memcached session resumption, session ticket synchronization using memcached / redis.

> we can benefit from binding close-exec and non-block flags with single call under linux.

The `accept` call to the socket that is accepting HTTP connections is made in `h2o_evloop_socket_accept`. The function can do that and already does that. What we need to make sure is that we do not set the flag twice when the function is called.

File: examples/libh2o/latency-optimization.c [hunk 1]
```diff
@@ -354,7 +354,12 @@ int main(int argc, char **argv)
 
     if (mode_listen) {
         int fd, reuseaddr_flag = 1;
-        if ((fd = socket(AF_INET, SOCK_STREAM, 0)) == -1 ||
+        if (
+#ifdef __linux__
+            (fd = socket(AF_INET, SOCK_STREAM | SOCK_CLOEXEC | SOCK_NONBLOCK, 0)) == -1 ||
+#else
+            (fd = socket(AF_INET, SOCK_STREAM, 0)) == -1 ||
+#endif
```
Comment (kazuho) [inline examples/libh2o/latency-optimization.c]: I would appreciate it if you could revert the change to minimize the maintenance cost of the code.

Considering that a HTTP/2 connection is typically long-running and that latency optimization issues many syscalls than when set to off, I do not think we need to complicate things here.

Comment (chenbd) [reply examples/libh2o/latency-optimization.c]: reverted.

**Comment by kazuho** [comment]
Thank you for the changes. Merged to master.
